src/test/ui/struct-literal-variant-in-if.stderr

```diff
@@ -0,0 +1,69 @@
+error: struct literals are not allowed here
+  --> $DIR/struct-literal-variant-in-if.rs:14:13
+   |
+LL |     if x == E::I { field1: true, field2: 42 } {}
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+help: surround the struct literal with parenthesis
+   |
+LL |     if x == (E::I { field1: true, field2: 42 }) {}
+   |             ^                                 ^
+
+error: struct literals are not allowed here
+  --> $DIR/struct-literal-variant-in-if.rs:16:13
+   |
+LL |     if x == E::V { field: false } {}
+   |             ^^^^^^^^^^^^^^^^^^^^^
+help: surround the struct literal with parenthesis
+   |
+LL |     if x == (E::V { field: false }) {}
+   |             ^                     ^
+
+error: struct literals are not allowed here
+  --> $DIR/struct-literal-variant-in-if.rs:18:13
+   |
+LL |     if x == E::J { field: -42 } {}
+   |             ^^^^^^^^^^^^^^^^^^^
+help: surround the struct literal with parenthesis
+   |
+LL |     if x == (E::J { field: -42 }) {}
+   |             ^                   ^
+
+error: struct literals are not allowed here
+  --> $DIR/struct-literal-variant-in-if.rs:20:13
+   |
+LL |     if x == E::K { field: "" } {}
+   |             ^^^^^^^^^^^^^^^^^^
+help: surround the struct literal with parenthesis
+   |
+LL |     if x == (E::K { field: "" }) {}
+   |             ^                  ^
+
+error[E0423]: expected value, found struct variant `E::V`
+  --> $DIR/struct-literal-variant-in-if.rs:10:13
+   |
+LL |     if x == E::V { field } {}
+   |             ^^^^----------
+   |             |
+   |             help: surround the struct literal with parenthesis: `(E::V { field })`
+
+error[E0423]: expected value, found struct variant `E::V`
+  --> $DIR/struct-literal-variant-in-if.rs:12:13
+   |
+LL |     if x == E::V { field: field } {}
+   |             ^^^^-----------------
+   |             |
+   |             help: surround the struct literal with parenthesis: `(E::V { field: field })`
+
+error[E0308]: mismatched types
+  --> $DIR/struct-literal-variant-in-if.rs:22:20
+   |
+LL |     let y: usize = ();
+   |                    ^^ expected usize, found ()
+   |
+   = note: expected type `usize`
+              found type `()`
+
+error: aborting due to 7 previous errors
+
+Some errors have detailed explanations: E0308, E0423.
+For more information about an error, try `rustc --explain E0308`.
```

Comment:
For comparison of the output without all of these changes: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1f4816149ca6e13ecb97d71c99675fec

```
error: expected identifier, found keyword `true`
  --> src/main.rs:14:28
   |
14 |     if x == E::I { field1: true, field2: 42 } {}
   |                            ^^^^ expected identifier, found keyword
help: you can escape reserved keywords to use them as identifiers
   |
14 |     if x == E::I { field1: r#true, field2: 42 } {}
   |                            ^^^^^^

error: expected type, found keyword `true`
  --> src/main.rs:14:28
   |
14 |     if x == E::I { field1: true, field2: 42 } {}
   |                            ^^^^ expecting a type here because of type ascription

error: expected identifier, found keyword `false`
  --> src/main.rs:16:27
   |
16 |     if x == E::V { field: false } {}
   |                           ^^^^^ expected identifier, found keyword
help: you can escape reserved keywords to use them as identifiers
   |
16 |     if x == E::V { field: r#false } {}
   |                           ^^^^^^^

error: expected type, found keyword `false`
  --> src/main.rs:16:27
   |
16 |     if x == E::V { field: false } {}
   |                           ^^^^^ expecting a type here because of type ascription

error: expected type, found `-`
  --> src/main.rs:18:27
   |
18 |     if x == E::J { field: -42 } {}
   |                           ^ expecting a type here because of type ascription

error: expected type, found `""`
  --> src/main.rs:20:27
   |
20 |     if x == E::K { field: "" } {}
   |                           ^^ expecting a type here because of type ascription

error[E0423]: expected value, found struct variant `E::V`
  --> src/main.rs:10:13
   |
10 |     if x == E::V { field } {}
   |             ^^^^ did you mean `E::V { /* fields */ }`?

error[E0423]: expected value, found struct variant `E::V`
  --> src/main.rs:12:13
   |
12 |     if x == E::V { field: field } {}
   |             ^^^^ did you mean `E::V { /* fields */ }`?

error[E0573]: expected type, found local variable `field`
  --> src/main.rs:12:27
   |
12 |     if x == E::V { field: field } {}
   |                           ^^^^^
   |                           |
   |                           not a type
   |                           expecting a type here because of type ascription

error[E0423]: expected value, found struct variant `E::I`
  --> src/main.rs:14:13
   |
14 |     if x == E::I { field1: true, field2: 42 } {}
   |             ^^^^ did you mean `E::I { /* fields */ }`?

error[E0423]: expected value, found struct variant `E::V`
  --> src/main.rs:16:13
   |
16 |     if x == E::V { field: false } {}
   |             ^^^^ did you mean `E::V { /* fields */ }`?

error[E0423]: expected value, found struct variant `E::J`
  --> src/main.rs:18:13
   |
18 |     if x == E::J { field: -42 } {}
   |             ^^^^ did you mean `E::J { /* fields */ }`?

error[E0423]: expected value, found struct variant `E::K`
  --> src/main.rs:20:13
   |
20 |     if x == E::K { field: "" } {}
   |             ^^^^ did you mean `E::K { /* fields */ }`?

error[E0658]: type ascription is experimental (see issue #23416)
  --> src/main.rs:12:20
   |
12 |     if x == E::V { field: field } {}
   |                    ^^^^^^^^^^^^

error[E0308]: mismatched types
  --> src/main.rs:10:20
   |
8  | fn test_E(x: E) {
   |                 - help: try adding a return type: `-> bool`
9  |     let field = true;
10 |     if x == E::V { field } {}
   |                    ^^^^^ expected (), found bool
   |
   = note: expected type `()`
              found type `bool`

error[E0308]: mismatched types
  --> src/main.rs:22:20
   |
22 |     let y: usize = ();
   |                    ^^ expected usize, found ()
   |
   = note: expected type `usize`
              found type `()`

error: aborting due to 16 previous errors
```

Reply:
A number of these was already covered by the previous PR - https://github.com/rust-lang/rust/pull/59981.